mono/utils/mono-mmap.c

```diff
@@ -197,6 +200,30 @@ prot_from_flags (int flags)
 	return prot;
 }
 
+#if defined(__APPLE__)
+
+#define DARWIN_VERSION_MOJAVE 18
+
+static guint32
+get_darwin_version (void)
+{
+	static guint32 version;
+	static gboolean res_set;
+
+	if (!res_set) {
```

Comment:
I realize this feature is yet x86-only, but this does not look thread safe on ARM.
Please consider pthread_once for initialization instead of rolling our own, which is surprisingly non trivial to get correct.

Reply:
This is actually thread safe, we are storing to 'version' which is a pointer and pointer sized writes are atomic even on arm.

Reply:
The specific missing thread safety is that barriers typically must come in pairs to have effect. The reader of res_set == TRUE, can still see uninitialized version.

Reply:
The atomicity of the write does not make this thread safe. The reader can see uninitialized data. The reader can see version==0 despite res_set==TRUE.

Reply:
The problem is that version can be read before res_set. The compiler and/or processor can reorder the reads.

The barrier on the write does not inhibit reordering of the reads.

One solution is a barrier between reading res_set and reading version.
Another solution is let pthread_once handle everything.
Another solution is probably to merge res_seat and version into one variable, i.e. "version 0" or (-1 or some other value) be the uninitialized value, instead of having the separate variable.

Reply:
So this doesn't seem thread safe on x86 either.